magma/magma
# chore(agw): Python dependency upgrades

Fixes #13426

In pairing with @wolfseb 

## Summary

Done:
- Upgraded Python versions where possible. We used `pipdeptree` to analyze packages and dependencies.
- Temporary change to upgrade `fpm` during provisioning. This is necessary to build `cryptography`.
- Some packages (such as docker) are not built by `pydep`, since a `python3-docker` exists in `http://archive.ubuntu.com/ubuntu/`.
- Updated lockfile


## Test Plan

-  Run `fab release package:destroy_vm=True` on the host VM to trigger new magma and dependency `*.deb` files to be built.
- Install local `*.deb` files, and any other dependencies that have not been newly packaged from the artifactory.
- Remove `/etc/magma`.
- Install `magma*.deb` file.
- Spin up `magma_test` and `magma_trfserver` 
- Run S1AP precommit and integration tests.
- Run `build-all` script on a fork: https://github.com/MoritzThomasHuebner/magma/runs/7759753398?check_suite_focus=true
  - For this run we needed to make a temporary change that disables the check that we are on the origin repo.
  - The check shows that all packages are actually being built.
- Run Bazel build and test workflow on fork: https://github.com/MoritzThomasHuebner/magma/actions/runs/2830383274
  - Ensures versions in `bazel/external/requirements.in` do not cause any version conflicts

## Additional Information

There are a number of packages we were not able to upgrade:
- `fpm` requires that there is a `setup.py`, thus some of the newest versions of `pyparsing`, `jsonschema`, `pyroute2` can no longer be built since these packages now utilize other build methods.
- `setuptools`: The version `49.6.0` is hardcoded in some places.
- `docker`: Ubuntu focal is not compatible with `docker>=5`.
  - `requests`: The latest versions require `docker>=5`.
- `flask` is restricted to `1.1.4`.
  - This restricts `click`, `itsdangerous`, `Jinja2`, and `markupsafe`.
- `redis` is restricted to `3.5.3`.
- `cryptography` is restricted to `3.2.1` because newer versions cannot be built with `fpm`.
- `PyYAML` is restricted to `<6.0`
